### PR TITLE
Server side pagination with maintenance schedule system list page

### PR DESCRIPTION
**Pull request by cbbayburt**
Adds `ApiDataEndpoint` JS module and Java backend helper classes to make paginated data requests from the server. The module is used by a new table component called `AsyncTable`.

`AsyncTable` is used to redesign the maintenance schedule system list page to improve the UX:
 - Paginate and sort data at the server
 - Add a search field for server-side filtering
 - Show both assigned and non-assigned systems in the table
 - Add a column to show currently assigned schedule labels

## GUI diff

Before:
maint-systems-before

After:
maint-systems-after

## Documentation
- No documentation needed: Functionality unchanged

## Test coverage
- Unit tests were added

## Links

Fixes https://github.com/SUSE/spacewalk/issues/11895

## Changelogs

If you don't need a changelog check, please mark this checkbox:

- [ ] No changelog needed

If you uncheck the checkbox after the PR is created, you will need to re-run `changelog_test` (see below)

## Re-run a test

If you need to re-run a test, please mark the related checkbox, it will be unchecked automatically once it has re-run:

- [ ] Re-run test "changelog_test"
- [ ] Re-run test "backend_unittests_pgsql"
- [ ] Re-run test "java_lint_checkstyle"
- [ ] Re-run test "java_pgsql_tests"
- [ ] Re-run test "schema_migration_test_oracle"
- [ ] Re-run test "schema_migration_test_pgsql"
- [ ] Re-run test "susemanager_unittests"
- [ ] Re-run test "javascript_lint"
- [ ] Re-run test "spacecmd_unittests"

### CODE REVIEW

**Comment by cbbayburt** [comment]
@admd I need a performance-oriented review on this. Don't bother with all the frontend/JS stuff, the performance-critical parts are all in the last commit. Thanks!

File: web/html/src/components/table/AsyncDataHandler.js [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+      }
+    });
+  }
+
+  UNSAFE_componentWillReceiveProps(nextProps: Props) {
+    this.setState({
+        selectedItems: nextProps.selectedItems || [],
+        selectable: Boolean(nextProps.selectable)
+    });
+  };
```
Comment (cbosdo) [inline web/html/src/components/table/AsyncDataHandler.js:117]: I know it's a copy/paste from `TableDataHandler` but would be nice to get rid of the `UNSAFE_*` functions in that code

File: web/html/src/components/table/AsyncDataHandler.js [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+  itemsPerPage: number,
+  totalItems: number,
+  criteria?: string,
+  sortColumnKey: string | null,
+  sortDirection: number,
+  selectedItems: Array<any>,
+  selectable: boolean,
+  loading: boolean,
+  endpoint: ApiDataEndpoint
+};
```
Comment (cbosdo) [inline web/html/src/components/table/AsyncDataHandler.js:63]: Most of these are common with `TableDataHandler`. I would rather see the `TableDataHandler` code refactored to accept something like a `dataProvider` property pointing to another component responsible for getting the data either asynchronously with your code or using preprovided data as we already have.

File: java/code/src/com/suse/manager/webui/utils/gson/PagedDataResultJson.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+/**
+ * JSON wrapper class to hold paginated data
+ *
+ * @param <T> the type of the data items
+ */
+public class PagedDataResultJson<T> {
+
+    private List<T> items;
+    private long total;
```
[the remark this answers is not in the document]
Comment (cbbayburt) [reply java/code/src/com/suse/manager/webui/utils/gson/PagedDataResultJson.java:30]: We also need it to display the total number in the UI, and in this case, this is enough to calculate the next/prev pages

File: web/html/src/utils/data-endpoint.js [hunk 4]
```diff
@@ -0,0 +1,131 @@
+// @flow
+
+import {debounce} from 'lodash';
+import {get} from './network';
+
+import type {Cancelable} from "../utils/functions";
+
+declare var userPrefPageSize: number;
```
[the remark this answers is not in the document]
Comment (cbbayburt) [reply web/html/src/utils/data-endpoint.js:8]: Can you point me to an example page?

[the remark this answers is not in the document]
Comment (cbosdo) [reply web/html/src/utils/data-endpoint.js:8]: Here you go: https://github.com/uyuni-project/uyuni/blob/master/web/html/src/manager/virtualization/guests/list/guests-list.renderer.js

**Review by admd — APPROVED** [review]
Look Good @cbbayburt. Thank you.

**Review by hustodemon — APPROVED** [review]
Looks good to me (didn't review the last 2 commits, because I have absolutely no clue, what's going on there).

**Review by parlt91 — APPROVED** [review]
Only had another look at your last two commits.
I have to admit, it took me a while to figure out what's going on there.
But I might not have understood every single details, so I put extra
focus into testing the implementation.
I couldn't find any issues with the existing uses of the `Table` component
also the adaption for the `maintenance-windows` works great.
So no reason for me not to approve.
Great job. Thanks :+1:

File: web/html/src/utils/data-providers/page-control.js [hunk 5]
```diff
@@ -0,0 +1,29 @@
+// @flow
+
+declare var userPrefPageSize: number;
```
Comment (cbosdo) [inline web/html/src/utils/data-providers/page-control.js:3]: Again the ugly global... :nauseated_face:

File: web/html/src/utils/data-providers/simple-data-provider.js [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+type PagedData = {
+  items: Array<any>,
+  total: number
+};
+
+export default class SimpleDataProvider {
+  data: Array<any>;
+  identifier: (row: any) => any;
+  filter: ?(row: any, criteria: string) => boolean;
+  comparators: ?{column: (a: any, b: any, key: string, direction: number) => number};
```
Comment (cbosdo) [inline web/html/src/utils/data-providers/simple-data-provider.js:14]: Would be easier to read with a `ComparatorType`. And that type would be reused further down too

File: web/html/src/components/table/TableDataHandler.js [hunk 7]
```diff
@@ -18,16 +21,27 @@ type ChildrenArgsProps = {
...
+   *
+   * The data returned from the endpoint must be in paginated form as the following:
+   * ```
+   * {
+   *    items: [..],
+   *    total: TOTAL_ITEMS
+   * }
+   * ```
+   *
+   * See: utils/data-providers/paged-data-endpoint.js for async usage
```
Comment (cbosdo) [inline web/html/src/components/table/TableDataHandler.js:36]: I like that approach. Thanks for the hard work on it :clap:

**Review by cbosdo — APPROVED** [review]
LGTM. Have you played with legacy tables to ensure it doesn't bring any regression?

File: java/code/src/com/suse/manager/webui/utils/test/PageControlHelperTest.java [hunk 8]
```diff
@@ -0,0 +1,289 @@
...
+        // Filter string and column
+        queryParams.put("q", "ang");
+        queryParams.put("qc", "secondProperty");
+
+        PageControlHelper helper = new PageControlHelper(SparkTestUtils.createMockRequestWithParams(REQUEST_URL,
+                queryParams), "mydefaultproperty");
+
+        DataResult<PagedDataItem> result = helper.processPageControl(testData);
+
+        assertEquals(2, result.size());
```
[the remark this answers is not in the document]
Comment (cbbayburt) [reply java/code/src/com/suse/manager/webui/utils/test/PageControlHelperTest.java:282]: No, the parameter `p` is actually for the start index (counting from 1), not the actual page number.

File: java/code/src/com/suse/manager/webui/controllers/maintenance/MaintenanceScheduleController.java [hunk 9]
```diff
@@ -255,11 +266,112 @@ public static String delete(Request request, Response response, User user) {
...
+
+        DataResult<SystemScheduleDto> systems = SystemManager.systemListWithSchedules(user, null);
+
+        if ("id".equals(pageHelper.getFunction())) {
+            // Return only IDs for "select all" function
+            return json(response, systems.stream().map(SystemScheduleDto::getId).collect(Collectors.toList()));
+        }
+
+        pageHelper.applySort(systems);
+        systems = pageHelper.processPageControl(systems, new HashMap<>());
```
[the remark this answers is not in the document]
Comment (cbbayburt) [reply java/code/src/com/suse/manager/webui/controllers/maintenance/MaintenanceScheduleController.java:311]: Right, I guess I had a reason for this in the early implementation, but it doesn't seem to be valid anymore. Thanks!

File: web/html/src/utils/data-providers/simple-data-provider.js (hunk 6 above)
Comment (cbbayburt) [reply web/html/src/utils/data-providers/simple-data-provider.js]: :+1:

File: web/html/src/utils/data-providers/page-control.js (hunk 5 above)
Comment (cbbayburt) [reply web/html/src/utils/data-providers/page-control.js]: Now I'm using the mixin provided from jade :+1:

**Pull request merged**